Reland: Inspector agents should use CheckedRef for InspectorEnvironment references.

basuke opened the pull request:
#### 0f0653d9eddc1300c8884b1033a5769b9cb06545
<pre>
Reland: Inspector agents should use CheckedRef for InspectorEnvironment references.
<a href="https://bugs.webkit.org/show_bug.cgi?id=302416">https://bugs.webkit.org/show_bug.cgi?id=302416</a>
<a href="https://rdar.apple.com/165006365">rdar://165006365</a>

Reviewed by Nobody (OOPS).

Inspector agents hold references to their InspectorEnvironment through the AgentContext
structure. This changes the raw reference to CheckedRef and WeakRef to provide compile
time and runtime lifetime checking.

The classes that had InspectorEnvironment reference are changed to store it as WeakRef.
When using that, checkedEnvironment() is used to ensure the validity of address during
its lifecycle.

Because XxxInspectorController classes are the concrete classes of InspectorEnvironment,
they are changed to be CheckedPtr-ready and WeakPtr-ready classes.

* Source/JavaScriptCore/API/APIUtils.h:
(handleExceptionIfNeeded):
(setException):
* Source/JavaScriptCore/API/JSBase.cpp:
(JSEvaluateScriptInternal):
(JSCheckScriptSyntax):
* Source/JavaScriptCore/API/JSValue.mm:
(reportExceptionToInspector):
* Source/JavaScriptCore/inspector/InjectedScript.cpp:
(Inspector::InjectedScript::execute):
(Inspector::InjectedScript::evaluate):
(Inspector::InjectedScript::awaitPromise):
(Inspector::InjectedScript::callFunctionOn):
(Inspector::InjectedScript::evaluateOnCallFrame):
(Inspector::InjectedScript::getFunctionDetails):
(Inspector::InjectedScript::functionDetails):
(Inspector::InjectedScript::getPreview):
(Inspector::InjectedScript::getProperties):
(Inspector::InjectedScript::getDisplayableProperties):
(Inspector::InjectedScript::getInternalProperties):
(Inspector::InjectedScript::getCollectionEntries):
(Inspector::InjectedScript::saveResult):
(Inspector::InjectedScript::wrapCallFrames const):
(Inspector::InjectedScript::wrapObject const):
(Inspector::InjectedScript::wrapJSONString const):
(Inspector::InjectedScript::wrapTable const):
(Inspector::InjectedScript::previewValue const):
(Inspector::InjectedScript::setEventValue):
(Inspector::InjectedScript::clearEventValue):
(Inspector::InjectedScript::setExceptionValue):
(Inspector::InjectedScript::clearExceptionValue):
(Inspector::InjectedScript::findObjectById const):
(Inspector::InjectedScript::inspectObject):
(Inspector::InjectedScript::releaseObject):
(Inspector::InjectedScript::releaseObjectGroup):
(Inspector::InjectedScript::createCommandLineAPIObject const):
* Source/JavaScriptCore/inspector/InjectedScriptBase.cpp:
(Inspector::InjectedScriptBase::hasAccessToInspectedScriptState const):
* Source/JavaScriptCore/inspector/InjectedScriptBase.h:
(Inspector::InjectedScriptBase::inspectorEnvironment const):
(Inspector::InjectedScriptBase::checkedInspectorEnvironment const):
* Source/JavaScriptCore/inspector/InjectedScriptManager.cpp:
(Inspector::InjectedScriptManager::InjectedScriptManager):
(Inspector::InjectedScriptManager::injectedScriptFor):
* Source/JavaScriptCore/inspector/InjectedScriptManager.h:
(Inspector::InjectedScriptManager::inspectorEnvironment const):
(Inspector::InjectedScriptManager::checkedInspectorEnvironment const):
* Source/JavaScriptCore/inspector/InjectedScriptModule.cpp:
(Inspector::InjectedScriptModule::ensureInjected):
* Source/JavaScriptCore/inspector/InspectorAgentBase.h:
* Source/JavaScriptCore/inspector/InspectorEnvironment.h:
* Source/JavaScriptCore/inspector/JSGlobalObjectInspectorController.h:
* Source/JavaScriptCore/inspector/agents/InspectorAgent.cpp:
(Inspector::InspectorAgent::initialized):
* Source/JavaScriptCore/inspector/agents/InspectorAgent.h:
* Source/JavaScriptCore/inspector/agents/InspectorAuditAgent.cpp:
(Inspector::InspectorAuditAgent::InspectorAuditAgent):
* Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp:
(Inspector::InspectorDebuggerAgent::InspectorDebuggerAgent):
(Inspector::InspectorDebuggerAgent::didPause):
(Inspector::InspectorDebuggerAgent::breakpointActionProbe):
(Inspector::InspectorDebuggerAgent::didContinue):
* Source/JavaScriptCore/inspector/agents/InspectorHeapAgent.cpp:
(Inspector::InspectorHeapAgent::enable):
(Inspector::InspectorHeapAgent::disable):
(Inspector::InspectorHeapAgent::gc):
(Inspector::InspectorHeapAgent::snapshot):
(Inspector::InspectorHeapAgent::nodeForHeapObjectIdentifier):
(Inspector::InspectorHeapAgent::getPreview):
(Inspector::InspectorHeapAgent::getRemoteObject):
(Inspector::InspectorHeapAgent::willGarbageCollect):
(Inspector::InspectorHeapAgent::didGarbageCollect):
(Inspector::InspectorHeapAgent::heapSnapshotBuilderIgnoreNode):
(Inspector::InspectorHeapAgent::clearHeapSnapshots):
* Source/JavaScriptCore/inspector/agents/InspectorHeapAgent.h:
* Source/JavaScriptCore/inspector/agents/InspectorRuntimeAgent.cpp:
(Inspector::InspectorRuntimeAgent::InspectorRuntimeAgent):
(Inspector::m_vm):
* Source/JavaScriptCore/inspector/agents/InspectorScriptProfilerAgent.cpp:
(Inspector::InspectorScriptProfilerAgent::willDestroyFrontendAndBackend):
(Inspector::InspectorScriptProfilerAgent::startTracking):
(Inspector::InspectorScriptProfilerAgent::stopTracking):
(Inspector::InspectorScriptProfilerAgent::willEvaluateScript):
(Inspector::InspectorScriptProfilerAgent::didEvaluateScript):
(Inspector::InspectorScriptProfilerAgent::trackingComplete):
(Inspector::InspectorScriptProfilerAgent::stopSamplingWhenDisconnecting):
* Source/JavaScriptCore/inspector/agents/InspectorScriptProfilerAgent.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::~JSGlobalObject):
(JSC::JSGlobalObject::init):
(JSC::JSGlobalObject::inspectorController const):
(JSC::JSGlobalObject::checkedInspectorController const):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::inspectorController const): Deleted.
* Source/JavaScriptCore/runtime/JSGlobalObjectDebuggable.cpp:
(JSC::JSGlobalObjectDebuggable::connect):
(JSC::JSGlobalObjectDebuggable::disconnect):
(JSC::JSGlobalObjectDebuggable::dispatchMessageFromRemote):
* Source/WebCore/inspector/FrameInspectorController.h:
* Source/WebCore/inspector/InspectorInstrumentation.cpp:
(WebCore::InspectorInstrumentation::frameWindowDiscardedImpl):
(WebCore::InspectorInstrumentation::didReceiveResourceResponseImpl):
(WebCore::InspectorInstrumentation::didFailLoadingImpl):
(WebCore::InspectorInstrumentation::didCommitLoadImpl):
(WebCore::InspectorInstrumentation::addMessageToConsoleImpl):
(WebCore::InspectorInstrumentation::consoleCountImpl):
(WebCore::InspectorInstrumentation::consoleCountResetImpl):
(WebCore::InspectorInstrumentation::startConsoleTimingImpl):
(WebCore::InspectorInstrumentation::logConsoleTimingImpl):
(WebCore::InspectorInstrumentation::stopConsoleTimingImpl):
* Source/WebCore/inspector/InspectorWebAgentBase.h:
(WebCore::InspectorAgentBase::checkedEnvironment):
* Source/WebCore/inspector/InstrumentingAgents.cpp:
(WebCore::InstrumentingAgents::developerExtrasEnabled const):
* Source/WebCore/inspector/InstrumentingAgents.h:
(WebCore::InstrumentingAgents::checkedEnvironment const):
(WebCore::InstrumentingAgents::inspectorEnvironment const): Deleted.
* Source/WebCore/inspector/PageInspectorController.h:
* Source/WebCore/inspector/WorkerInspectorController.h:
* Source/WebCore/inspector/agents/InspectorAnimationAgent.cpp:
(WebCore::InspectorAnimationAgent::startTracking):
(WebCore::InspectorAnimationAgent::stopTracking):
(WebCore::InspectorAnimationAgent::willApplyKeyframeEffect):
(WebCore::InspectorAnimationAgent::stopTrackingStyleOriginatedAnimation):
* Source/WebCore/inspector/agents/InspectorCPUProfilerAgent.cpp:
(WebCore::InspectorCPUProfilerAgent::startTracking):
(WebCore::InspectorCPUProfilerAgent::stopTracking):
(WebCore::InspectorCPUProfilerAgent::collectSample):
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
(WebCore::InspectorDOMAgent::mediaMetricsTimerFired):
* Source/WebCore/inspector/agents/InspectorMemoryAgent.cpp:
(WebCore::InspectorMemoryAgent::startTracking):
(WebCore::InspectorMemoryAgent::stopTracking):
(WebCore::InspectorMemoryAgent::didHandleMemoryPressure):
(WebCore::InspectorMemoryAgent::collectSample):
* Source/WebCore/inspector/agents/InspectorNetworkAgent.cpp:
(WebCore::InspectorNetworkAgent::buildObjectForTiming):
(WebCore::InspectorNetworkAgent::timestamp):
(WebCore::InspectorNetworkAgent::didFinishLoading):
* Source/WebCore/inspector/agents/InspectorPageAgent.cpp:
(WebCore::InspectorPageAgent::enable):
(WebCore::InspectorPageAgent::timestamp):
* Source/WebCore/inspector/agents/InspectorTimelineAgent.cpp:
(WebCore::InspectorTimelineAgent::internalStart):
(WebCore::InspectorTimelineAgent::internalStop):
(WebCore::InspectorTimelineAgent::timestamp):
(WebCore::InspectorTimelineAgent::timestampFromMonotonicTime):
* Source/WebCore/inspector/agents/page/PageTimelineAgent.cpp:
(WebCore::PageTimelineAgent::internalStart):
(WebCore::PageTimelineAgent::didCompleteRenderingFrame):
* Source/WebKitLegacy/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebKitLegacy/mac/WebInspector/WebNodeHighlightView.mm:
(-[WebNodeHighlightView drawRect:]):

Canonical link: <a href="https://commits.webkit.org/304015@main">https://commits.webkit.org/304015@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4b03fb7935d089a3a0dc71367683f58bf4ebe682

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134027 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6538 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45232 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141607 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86089 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/322c469a-ca2f-4cda-b89e-eb3a007bb90a) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/135897 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7071 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6402 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102548 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/69820 "Passed tests") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/1f0d3b8c-8749-43f3-82d5-57b0b6b43ea2) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136974 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/4960 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120157 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83344 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/34efa385-1990-48cf-9693-9404a3e280ad) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4838 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2465 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/1422 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/126104 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/114010 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38278 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144253 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/132541 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6208 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38855 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110895 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6290 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5260 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111113 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28238 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4703 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116414 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/59978 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6260 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34659 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/165504 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6106 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69724 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43207 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6351 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6214 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

burg commented on Source/JavaScriptCore/inspector/agents/InspectorScriptProfilerAgent.cpp:
```diff
@@ -122,22 +128,30 @@ Seconds InspectorScriptProfilerAgent::willEvaluateScript()
     m_activeEvaluateScript = true;
 
+    CheckedPtr environment = m_environment.get();
+    if (!environment)
```
This should be makeUnexpected to match the other case where the environment is gone, i think?

burg commented on Source/WebCore/inspector/InspectorController.h:
```diff
@@ -64,13 +65,21 @@ class PageDebugger;
 class WebInjectedScriptManager;
 struct PageAgentContext;
 
-class InspectorController final : public Inspector::InspectorEnvironment, public CanMakeWeakPtr<InspectorController> {
```
You'll need to rebase this PR since this class got renamed to PageInspectorController.

burg commented on Source/WebCore/inspector/agents/InspectorTimelineAgent.cpp:
```diff
@@ -178,7 +178,8 @@ void InspectorTimelineAgent::internalStart(std::optional<int>&& maxCallStackDept
     Ref { m_instrumentingAgents.get() }->setTrackingTimelineAgent(this);
 
-    m_environment.debugger()->addObserver(*this);
+    if (CheckedPtr environment = m_environment.get())
```
Furthering my comment above, this is a good case where we should probably just bail out immediately `if (!m_environment)` at the top of the method since we don't expect it to work if we can't access `environment->debugger()`.

burg commented on Source/WebCore/inspector/agents/InspectorMemoryAgent.cpp:
```diff
@@ -164,12 +168,14 @@ void InspectorMemoryAgent::collectSample(const ResourceUsageData& data)
-        .setTimestamp(m_environment.executionStopwatch().elapsedTimeSince(data.timestamp).seconds())
-        .setCategories(WTFMove(categories))
-        .release();
+    if (CheckedPtr environment = m_environment.get()) {
```
We don't expect environment to be null. So, in most cases, I prefer using early exit `if (!environment) return { }` or similar, because it reduces code indentation and it can be turned into a macro easily if we need to perform multiple WeakPtr validations. See macros in `WebAutomationSession.h`.

the-chenergy commented on Source/JavaScriptCore/inspector/agents/InspectorHeapAgent.h:
```diff
@@ -83,7 +83,7 @@ class JS_EXPORT_PRIVATE InspectorHeapAgent : public InspectorAgentBase, public H
     const UniqueRef<HeapFrontendDispatcher> m_frontendDispatcher;
     const Ref<HeapBackendDispatcher> m_backendDispatcher;
-    InspectorEnvironment& m_environment;
+    WeakPtr<InspectorEnvironment> m_environment;
```
Rather than changing a `&` to `WeakPtr`, do you think it's worthwhile to find out why it's possible for the agent to outlive the environment _without being notified?_ It seems like our design is indeed to notify the agent through `willDestroyFrontendAndBackend` that the environment is gone. Perhaps that doesn't always happen in time?

Could it be because the ordering of `unregisterInstrumentingAgents` and `willDestroyFrontendAndBackend` differs [here,](https://github.com/WebKit/WebKit/blob/4291549ea228120beabdd904cacb95428665d20f/Source/WebCore/inspector/PageInspectorController.cpp#L325-L329) in contrast to [here?](https://github.com/WebKit/WebKit/blob/4291549ea228120beabdd904cacb95428665d20f/Source/WebCore/inspector/PageInspectorController.cpp#L293-L299)

basuke replied:
If every objects are managed by C++ object, the design should work well. In reality, some objects are managed by GC world which can delay actual disposal a bit later. That timing difference may cause the object that shouldn't live longer than other referencing object get accessed after free.

In the original case, `InstrumentingAgents` holds the reference to `InspectorEnvironment`. Because `InstrumentingAgents` is managed by GC, it lives longer than `InspectorEnvironment` (usually this is same with `InspectorController` that is hold by Page). That means the reference (or CheckedPtr) to `InspectorEnvironment` get stored in `InstrumentingAgents` even if they are actually destroyed. This is a bad situation. This is what Safer CPP tries to solve.

the-chenergy replied:
Right, and your point was to rely on `WeakPtr` automatically becomes null after the underlying object (the `InspectorEnvironment`) gets destroyed, right? Given you used a `CheckedPtr` [here,](https://github.com/WebKit/WebKit/pull/54147/files#diff-963d93278ee4f843142e722ef5bb420fddf303b62af47547c9c60c72beffd728R67), if the environment gets destroyed right after the null check, it'll still crash although without security implications. Shouldn't the proper fix of the UAF be to make sure `InspectorHeapAgent::disconnect` gets called timely, before the environment gets destroyed whose timing we should have knowledge about?

dcrousso replied:
to be clear, `InstrumentingAgents` is not managed by GC.  it is a dynamically allocated and refcounted object.  it will only stay alive so long as there are things that reference it and therefore keep it alive

`InstrumentingAgents` is created by the `InspectorEnvironment` subclass (e.g. `PageInspectorController`, `WorkerInspectorController`, etc.), and IMO the fact that it uses a `Ref` (instead of `CheckedRef`) is a mistake because it should never be the case that the former outlives the latter

i agree with @the-chenergy that we should be finding out how that's possible and fixing it to not do that instead of adjusting the behavior of everything else (i.e. all of the `InspectorAgent` currently rightly assume that the `InspectorEnvironment` will always exist and that is not something we want to change)

basuke replied:
> Given you used a CheckedPtr here, if the environment gets destroyed right after the null check, it'll still crash although without security implications.

Sorry I don't get which line you mentioned. Assume that is in the middle of the function after checked point is checked and later in the same function, if that the case we need to worry, then we should use RefCounted instead of CheckedPtr. I think it is safe to check only at the function top because of the design. The original value was just a ref. That means the usage expects the object alive during its lifecycle. That's why it was changed by simple ref to CheckedRef for the sake of Safer CPP. With this change, actual crash happened.

Actual crash is like this:

```
VM Region Info: 0x18 is not in any region.  Bytes before following region: 4307238888
      REGION TYPE                    START - END         [ VSIZE] PRT/MAX SHRMOD  REGION DETAIL
      UNUSED SPACE AT START
--->  
      __TEXT                      100bb4000-100c58000    [  656K] r-x/r-x SM=COW  /Volumes/VOLUME/*/DumpRenderTree

Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   WebCore                       	       0x11bc0a5ac WTF::CheckedRef<Inspector::InspectorEnvironment, WTF::RawPtrTraits<Inspector::InspectorEnvironment>>::~CheckedRef() + 16 (CheckedRef.h:56) [inlined]
1   WebCore                       	       0x11bc0a5ac WTF::CheckedRef<Inspector::InspectorEnvironment, WTF::RawPtrTraits<Inspector::InspectorEnvironment>>::~CheckedRef() + 16 (CheckedRef.h:53) [inlined]
2   WebCore                       	       0x11bc0a5ac WebCore::InstrumentingAgents::~InstrumentingAgents() + 60 (InstrumentingAgents.h:149)
3   WebCore                       	       0x11bc007bc WebCore::InstrumentingAgents::~InstrumentingAgents() + 4 (InstrumentingAgents.h:149) [inlined]
4   WebCore                       	       0x11bc007bc WTF::RefCounted<WebCore::InstrumentingAgents>::deref() const + 24 (RefCounted.h:200) [inlined]
5   WebCore                       	       0x11bc007bc WTF::RefCountedAndCanMakeWeakPtr<WebCore::InstrumentingAgents>::deref() const + 24 (RefCountedAndCanMakeWeakPtr.h:37) [inlined]
6   WebCore                       	       0x11bc007bc WTF::DefaultRefDerefTraits<WebCore::InstrumentingAgents>::derefIfNotNull(WebCore::InstrumentingAgents*) + 28 (Ref.h:64) [inlined]
7   WebCore                       	       0x11bc007bc WTF::RefPtr<WebCore::InstrumentingAgents, WTF::RawPtrTraits<WebCore::InstrumentingAgents>, WTF::DefaultRefDerefTraits<WebCore::InstrumentingAgents>>::~RefPtr() + 36 (RefPtr.h:62) [inlined]
8   WebCore                       	       0x11bc007bc WTF::RefPtr<WebCore::InstrumentingAgents, WTF::RawPtrTraits<WebCore::InstrumentingAgents>, WTF::DefaultRefDerefTraits<WebCore::InstrumentingAgents>>::~RefPtr() + 36 (RefPtr.h:62) [inlined]
9   WebCore                       	       0x11bc007bc WebCore::CommandLineAPIHost::~CommandLineAPIHost() + 196 (CommandLineAPIHost.cpp:77)
10  WebCore                       	       0x11a332420 WTF::RefCounted<WebCore::CommandLineAPIHost>::deref() const + 4 (RefCounted.h:200) [inlined]
11  WebCore                       	       0x11a332420 WTF::DefaultRefDerefTraits<WebCore::CommandLineAPIHost>::derefIfNotNull(WebCore::CommandLineAPIHost*) + 4 (Ref.h:64) [inlined]
12  WebCore                       	       0x11a332420 WTF::Ref<WebCore::CommandLineAPIHost, WTF::RawPtrTraits<WebCore::CommandLineAPIHost>, WTF::DefaultRefDerefTraits<WebCore::CommandLineAPIHost>>::~Ref() + 4 (Ref.h:85) [inlined]
13  WebCore                       	       0x11a332420 WTF::Ref<WebCore::CommandLineAPIHost, WTF::RawPtrTraits<WebCore::CommandLineAPIHost>, WTF::DefaultRefDerefTraits<WebCore::CommandLineAPIHost>>::~Ref() + 4 (Ref.h:79) [inlined]
14  WebCore                       	       0x11a332420 WebCore::JSDOMWrapper<WebCore::CommandLineAPIHost, WTF::RawPtrTraits<WebCore::CommandLineAPIHost>>::~JSDOMWrapper() + 4 (JSDOMWrapper.h:75) [inlined]
15  WebCore                       	       0x11a332420 WebCore::JSCommandLineAPIHost::~JSCommandLineAPIHost() + 4 (JSCommandLineAPIHost.h:30) [inlined]
16  WebCore                       	       0x11a332420 WebCore::JSCommandLineAPIHost::~JSCommandLineAPIHost() + 4 (JSCommandLineAPIHost.h:30) [inlined]
17  WebCore                       	       0x11a332420 WebCore::JSCommandLineAPIHost::destroy(JSC::JSCell*) + 52 (JSCommandLineAPIHost.cpp:182)
```

`JSCommandLineAPIHost` is destroyed that was the last owner of the `InstrumentingAgents` and it was destroyed. Then its destructor catches the CheckedRef<InspectorEnvironment>. This means after `InspectorEnvironment` is actually destroyed and before `InstrumentingAgents` is destroyed, `InstrumentingAgents` has an invalid reference to `InspectorEnvironment`. This is what we want to solve.

basuke replied:
> to be clear, InstrumentingAgents is not managed by GC. it is a dynamically allocated and refcounted object. it will only stay alive so long as there are things that reference it and therefore keep it alive

Sorry, as I wrote above, `JSCommandLineAPIHost` is the GC managed object.

> IMO the fact that it uses a Ref (instead of CheckedRef) is a mistake because it should never be the case that the former outlives the latter

If being InstrumentingAgents as RefCounted is the mistake, we can fix that in the next patch. The focus of this PR is `InspectorEnvironment` and from that view, `InstrumentingAgents` lives longer that that. The PR is already too broad and I want to separate the issue.

> i agree with @the-chenergy that we should be finding out how that's possible and fixing it to not do that instead of adjusting the behavior of everything else (i.e. all of the InspectorAgent currently rightly assume that the InspectorEnvironment will always exist and that is not something we want to change)

That is the goal of Safer CPP activity. We need to do that. Of cause we can add `CheckedRef<InspectorEnvironment> checkedEnvironment()` member function to simplify the usage. Still we need the weakPtr implementation, but codebase is much simpler and existence of that method tells InspectorEnvironment exists during the object lifecycle. If it fails, it will be caught by crash report.

dcrousso replied:
> The focus of this PR is `InspectorEnvironment` and from that view, `InstrumentingAgents` lives longer that that.

this is a bug

we should be focusing on fixing this issue instead of covering it up with other changes

the expectation of all the `InspectorAgent` is that they will never outlive the `InspectorEnvironment` (especially since the latter creates the former)

for the purposes of Safer CPP it's absolutely fine to use a `CheckedRef<InspectorEnvironment>` instead of `InspectorEnvironment&` but we do not want to change anything else

as far as `CommandLineAPIHost`, there's already a `disconnect` method to solve this exact problem of `InstrumentingAgents` being kept alive, so it seems like there's some scenario where that is not being called when it should

basuke replied:
Okay, I hold this PR until I fix that bug. Let me clarify one thing. Should the `InstrumentingAgent` really be `RecCounted`?

the-chenergy replied:
I think the answer is no, given @dcrousso's [earlier comment.](https://github.com/WebKit/WebKit/pull/54147#discussion_r2543622216) The patch that made `InstrumentingAgents` ref-counted [(125490@main)](https://commits.webkit.org/125490@main) doesn't provide a strong argument because according to our current discussion, the users of `InstrumentingAgents` should be notified to disconnect in time and should therefore use `Checked{Ref, Ptr}` instead.

burg replied:
Somehow in this discussion, I missed that `InspectorInstrumentationCookie` kept a strong reference to `InstrumentingAgents`. It seems like an obvious case to deploy WeakPtr, right?

But then I looked at how old this code is, and this was done at a time when WeakPtr did not yet exist (125924@main).

> for the purposes of Safer CPP it's absolutely fine to use a CheckedRef<InspectorEnvironment> instead of InspectorEnvironment& but we do not want to change anything else

Let's focus this PR on the proposed refactoring. Hopefully we can deal with un-refcounting InstrumentingAgents in a separate PR.

dcrousso reviewed (changes requested):
please see <https://github.com/WebKit/WebKit/pull/54147#discussion_r2543622216>

burg reviewed (commented):
I think this addresses all of the outstanding feedback and doesn't change the lifetime of anything.

However, the Changelog is outdated, and the PR discussion is getting long. Please update the changelog to reflect the design that we landed on.

It would be good to call out and explain the weird cases in the changelog, such as `InspectorEnvironment* -> WeakPtr<InspectorEnvironment>` in InjectedScriptBase, and why its not CheckedPtr. Same goes for `InspectorEnvironment& -> WeakPtr<InpectorEnvironment>` in InjectedScriptManager.

burg commented on Source/JavaScriptCore/inspector/InspectorAgentBase.h:
```diff
@@ -84,4 +84,18 @@ class InspectorAgentBase {
     String m_name;
 };
 
+class InspectorEnvironmentHolder {
```
Two pieces of feedback: I don't like the name, and it's unclear why this is implemented using multiple inheritance instead of a separate struct. Maybe `InspectorEnvironmentClient` or `InspectorEnvironmentUsing` would better convey why some agents subclass it.

dcrousso replied:
yeah it's kinda odd to me as well that this even exists in the first place.  it seems simple enough that it should just be part of the things that inherit from it instead of as a separate class

burg reviewed (approved):
r+ but please include new changelog details to be proofread before landing. I'd also like Devin to take another look.

dcrousso commented on Source/JavaScriptCore/inspector/InspectorAgentBase.h:
```diff
@@ -84,4 +84,25 @@ class InspectorAgentBase {
+    InspectorEnvironment* environment() { return m_environment.get(); }
+    const InspectorEnvironment* environment() const { return m_environment.get(); }
+    CheckedRef<InspectorEnvironment> checkedEnvironment() { return *environment(); }
+    CheckedRef<const InspectorEnvironment> checkedEnvironment() const { return *environment(); }
```
it seems very odd that we convert from a `WeakRef` to a `CheckedRef`.  is that normal?  i thought the idea was that we'd always only use one or the other

basuke replied:
Yes, this is normal behavior. WeakRef validates the address only at the point of use. CheckedRef, on the other hand, tracks the address's lifecycle, allowing Safer C++ to detect issues at compile time.

dcrousso commented on Source/WebCore/inspector/agents/InspectorMemoryAgent.cpp:
```diff
@@ -99,7 +99,7 @@ Inspector::Protocol::ErrorStringOr<void> InspectorMemoryAgent::startTracking()
     m_tracking = true;
 
-    m_frontendDispatcher->trackingStart(m_environment.executionStopwatch().elapsedTime().seconds());
+        m_frontendDispatcher->trackingStart(checkedEnvironment()->executionStopwatch().elapsedTime().seconds());
```
oops?

dcrousso commented on Source/JavaScriptCore/inspector/InspectorAgentBase.h:
```diff
@@ -84,4 +84,25 @@ class InspectorAgentBase {
+    CheckedRef<const InspectorEnvironment> checkedEnvironment() const { return *environment(); }
+
+private:
+    WeakPtr<InspectorEnvironment> m_environment;
```
this really just feels like you've moved where the member variable is and tried to work around my objection by wrapping this in member functions that use `CheckedRef`

my previous <https://github.com/WebKit/WebKit/pull/54147#discussion_r2544002545> still stands

all of the inspector agents should be able to verifiably know that they will never outlive the `InspectorEnvironment` as otherwise that is a bug

at the very least id expect this to be a `WeakRef`, but again i think that this being any form of "weak" is really incorrect since we know/expect that the `InspectorEnvironment` will outlive all agents

aproskuryakov replied:
Devin, are you saying that this PR makes things objectively worse, or that it's not as good as it could be? If it's the latter, we should land ASAP, and work on further improvement separately. It would not be OK to keep a crasher regression in the tree just because the fix could be even nicer.

basuke replied:
As I said before, there's a case that assumption is wrong. If that is a bug, we need to keep it as `WeakPtr` (or `WeakRef`). Once we fix that bug, then I agree to change it as `CheckedRef`.

When we use environment, we need to the CheckedRef on stack to make sure the usage is checked by the Safer CPP. CheckedRef ensures the pointer is valid from its lifecycle in the function. So usage should be like this even if the member variable is stored using CheckedRef.

burg reviewed (approved):
Nice!

basuke committed:
Reland: Inspector agents should use CheckedRef for InspectorEnvironment references.
https://bugs.webkit.org/show_bug.cgi?id=302416
rdar://165006365

Reviewed by Nobody (OOPS).

Inspector agents hold references to their InspectorEnvironment through the AgentContext
structure. This changes the raw reference to CheckedRef and WeakRef to provide compile
time and runtime lifetime checking.

The classes that had InspectorEnvironment reference are changed to store it as WeakRef.
When using that, checkedEnvironment() is used to ensure the validity of address during
its lifecycle.

Because XxxInspectorController classes are the concrete classes of InspectorEnvironment,
they are changed to be CheckedPtr-ready and WeakPtr-ready classes.

* Source/JavaScriptCore/API/APIUtils.h:
(handleExceptionIfNeeded):
(setException):
* Source/JavaScriptCore/API/JSBase.cpp:
(JSEvaluateScriptInternal):
(JSCheckScriptSyntax):
* Source/JavaScriptCore/API/JSValue.mm:
(reportExceptionToInspector):
* Source/JavaScriptCore/inspector/InjectedScript.cpp:
(Inspector::InjectedScript::execute):
(Inspector::InjectedScript::evaluate):
(Inspector::InjectedScript::awaitPromise):
(Inspector::InjectedScript::callFunctionOn):
(Inspector::InjectedScript::evaluateOnCallFrame):
(Inspector::InjectedScript::getFunctionDetails):
(Inspector::InjectedScript::functionDetails):
(Inspector::InjectedScript::getPreview):
(Inspector::InjectedScript::getProperties):
(Inspector::InjectedScript::getDisplayableProperties):
(Inspector::InjectedScript::getInternalProperties):
(Inspector::InjectedScript::getCollectionEntries):
(Inspector::InjectedScript::saveResult):
(Inspector::InjectedScript::wrapCallFrames const):
(Inspector::InjectedScript::wrapObject const):
(Inspector::InjectedScript::wrapJSONString const):
(Inspector::InjectedScript::wrapTable const):
(Inspector::InjectedScript::previewValue const):
(Inspector::InjectedScript::setEventValue):
(Inspector::InjectedScript::clearEventValue):
(Inspector::InjectedScript::setExceptionValue):
(Inspector::InjectedScript::clearExceptionValue):
(Inspector::InjectedScript::findObjectById const):
(Inspector::InjectedScript::inspectObject):
(Inspector::InjectedScript::releaseObject):
(Inspector::InjectedScript::releaseObjectGroup):
(Inspector::InjectedScript::createCommandLineAPIObject const):
* Source/JavaScriptCore/inspector/InjectedScriptBase.cpp:
(Inspector::InjectedScriptBase::hasAccessToInspectedScriptState const):
* Source/JavaScriptCore/inspector/InjectedScriptBase.h:
(Inspector::InjectedScriptBase::inspectorEnvironment const):
(Inspector::InjectedScriptBase::checkedInspectorEnvironment const):
* Source/JavaScriptCore/inspector/InjectedScriptManager.cpp:
(Inspector::InjectedScriptManager::InjectedScriptManager):
(Inspector::InjectedScriptManager::injectedScriptFor):
* Source/JavaScriptCore/inspector/InjectedScriptManager.h:
(Inspector::InjectedScriptManager::inspectorEnvironment const):
(Inspector::InjectedScriptManager::checkedInspectorEnvironment const):
* Source/JavaScriptCore/inspector/InjectedScriptModule.cpp:
(Inspector::InjectedScriptModule::ensureInjected):
* Source/JavaScriptCore/inspector/InspectorAgentBase.h:
* Source/JavaScriptCore/inspector/InspectorEnvironment.h:
* Source/JavaScriptCore/inspector/JSGlobalObjectInspectorController.h:
* Source/JavaScriptCore/inspector/agents/InspectorAgent.cpp:
(Inspector::InspectorAgent::initialized):
* Source/JavaScriptCore/inspector/agents/InspectorAgent.h:
* Source/JavaScriptCore/inspector/agents/InspectorAuditAgent.cpp:
(Inspector::InspectorAuditAgent::InspectorAuditAgent):
* Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp:
(Inspector::InspectorDebuggerAgent::InspectorDebuggerAgent):
(Inspector::InspectorDebuggerAgent::didPause):
(Inspector::InspectorDebuggerAgent::breakpointActionProbe):
(Inspector::InspectorDebuggerAgent::didContinue):
* Source/JavaScriptCore/inspector/agents/InspectorHeapAgent.cpp:
(Inspector::InspectorHeapAgent::enable):
(Inspector::InspectorHeapAgent::disable):
(Inspector::InspectorHeapAgent::gc):
(Inspector::InspectorHeapAgent::snapshot):
(Inspector::InspectorHeapAgent::nodeForHeapObjectIdentifier):
(Inspector::InspectorHeapAgent::getPreview):
(Inspector::InspectorHeapAgent::getRemoteObject):
(Inspector::InspectorHeapAgent::willGarbageCollect):
(Inspector::InspectorHeapAgent::didGarbageCollect):
(Inspector::InspectorHeapAgent::heapSnapshotBuilderIgnoreNode):
(Inspector::InspectorHeapAgent::clearHeapSnapshots):
* Source/JavaScriptCore/inspector/agents/InspectorHeapAgent.h:
* Source/JavaScriptCore/inspector/agents/InspectorRuntimeAgent.cpp:
(Inspector::InspectorRuntimeAgent::InspectorRuntimeAgent):
(Inspector::m_vm):
* Source/JavaScriptCore/inspector/agents/InspectorScriptProfilerAgent.cpp:
(Inspector::InspectorScriptProfilerAgent::willDestroyFrontendAndBackend):
(Inspector::InspectorScriptProfilerAgent::startTracking):
(Inspector::InspectorScriptProfilerAgent::stopTracking):
(Inspector::InspectorScriptProfilerAgent::willEvaluateScript):
(Inspector::InspectorScriptProfilerAgent::didEvaluateScript):
(Inspector::InspectorScriptProfilerAgent::trackingComplete):
(Inspector::InspectorScriptProfilerAgent::stopSamplingWhenDisconnecting):
* Source/JavaScriptCore/inspector/agents/InspectorScriptProfilerAgent.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::~JSGlobalObject):
(JSC::JSGlobalObject::init):
(JSC::JSGlobalObject::inspectorController const):
(JSC::JSGlobalObject::checkedInspectorController const):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::inspectorController const): Deleted.
* Source/JavaScriptCore/runtime/JSGlobalObjectDebuggable.cpp:
(JSC::JSGlobalObjectDebuggable::connect):
(JSC::JSGlobalObjectDebuggable::disconnect):
(JSC::JSGlobalObjectDebuggable::dispatchMessageFromRemote):
* Source/WebCore/inspector/FrameInspectorController.h:
* Source/WebCore/inspector/InspectorInstrumentation.cpp:
(WebCore::InspectorInstrumentation::frameWindowDiscardedImpl):
(WebCore::InspectorInstrumentation::didReceiveResourceResponseImpl):
(WebCore::InspectorInstrumentation::didFailLoadingImpl):
(WebCore::InspectorInstrumentation::didCommitLoadImpl):
(WebCore::InspectorInstrumentation::addMessageToConsoleImpl):
(WebCore::InspectorInstrumentation::consoleCountImpl):
(WebCore::InspectorInstrumentation::consoleCountResetImpl):
(WebCore::InspectorInstrumentation::startConsoleTimingImpl):
(WebCore::InspectorInstrumentation::logConsoleTimingImpl):
(WebCore::InspectorInstrumentation::stopConsoleTimingImpl):
* Source/WebCore/inspector/InspectorWebAgentBase.h:
(WebCore::InspectorAgentBase::checkedEnvironment):
* Source/WebCore/inspector/InstrumentingAgents.cpp:
(WebCore::InstrumentingAgents::developerExtrasEnabled const):
* Source/WebCore/inspector/InstrumentingAgents.h:
(WebCore::InstrumentingAgents::checkedEnvironment const):
(WebCore::InstrumentingAgents::inspectorEnvironment const): Deleted.
* Source/WebCore/inspector/PageInspectorController.h:
* Source/WebCore/inspector/WorkerInspectorController.h:
* Source/WebCore/inspector/agents/InspectorAnimationAgent.cpp:
(WebCore::InspectorAnimationAgent::startTracking):
(WebCore::InspectorAnimationAgent::stopTracking):
(WebCore::InspectorAnimationAgent::willApplyKeyframeEffect):
(WebCore::InspectorAnimationAgent::stopTrackingStyleOriginatedAnimation):
* Source/WebCore/inspector/agents/InspectorCPUProfilerAgent.cpp:
(WebCore::InspectorCPUProfilerAgent::startTracking):
(WebCore::InspectorCPUProfilerAgent::stopTracking):
(WebCore::InspectorCPUProfilerAgent::collectSample):
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
(WebCore::InspectorDOMAgent::mediaMetricsTimerFired):
* Source/WebCore/inspector/agents/InspectorMemoryAgent.cpp:
(WebCore::InspectorMemoryAgent::startTracking):
(WebCore::InspectorMemoryAgent::stopTracking):
(WebCore::InspectorMemoryAgent::didHandleMemoryPressure):
(WebCore::InspectorMemoryAgent::collectSample):
* Source/WebCore/inspector/agents/InspectorNetworkAgent.cpp:
(WebCore::InspectorNetworkAgent::buildObjectForTiming):
(WebCore::InspectorNetworkAgent::timestamp):
(WebCore::InspectorNetworkAgent::didFinishLoading):
* Source/WebCore/inspector/agents/InspectorPageAgent.cpp:
(WebCore::InspectorPageAgent::enable):
(WebCore::InspectorPageAgent::timestamp):
* Source/WebCore/inspector/agents/InspectorTimelineAgent.cpp:
(WebCore::InspectorTimelineAgent::internalStart):
(WebCore::InspectorTimelineAgent::internalStop):
(WebCore::InspectorTimelineAgent::timestamp):
(WebCore::InspectorTimelineAgent::timestampFromMonotonicTime):
* Source/WebCore/inspector/agents/page/PageTimelineAgent.cpp:
(WebCore::PageTimelineAgent::internalStart):
(WebCore::PageTimelineAgent::didCompleteRenderingFrame):
* Source/WebKitLegacy/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebKitLegacy/mac/WebInspector/WebNodeHighlightView.mm:
(-[WebNodeHighlightView drawRect:]):

Canonical link: https://commits.webkit.org/304015@main
